gh-134262: Add retries to generate_sbom.py

emmatyping opened the pull request:
`Tools/build/generate_sbom.py` downloads artifacts, which can flake. This commit adds simple exponential back-off retries to the download to avoid failures due to network flakes.

<!-- gh-issue-number: gh-134262 -->
* Issue: gh-134262
<!-- /gh-issue-number -->

emmatyping committed:
Add retries to generate_sbom.py

generate_sbom.py downloads artifacts, which can
flake. This commit adds exponential backoff to the
download to avoid failures due to network flakes.

emmatyping committed:
Fix mypy for retry function

Ironic

sethmlarson reviewed (approved):
LGTM, thank you!

emmatyping committed:
Add Any return type

emmatyping committed:
Sort imports

donbarbos commented on Tools/build/generate_sbom.py:
```diff
@@ -161,6 +163,23 @@ def get_externals() -> list[str]:
 
+def download_with_retries(
+    download_location: str,
+    retries: int = 5) -> typing.Any:
```
I think we have a more general way to move arguments to a new line:
```suggestion
def download_with_retries(download_location: str,
                          retries: int = 5) -> typing.Any:
```
but I think that `Any` can be gotten rid of:
```suggestion
def download_with_retries(download_location: str, retries: int = 5):
```

donbarbos commented on Tools/build/generate_sbom.py:
```diff
@@ -161,6 +163,23 @@ def get_externals() -> list[str]:
+    """Download a file with exponential backoff retry."""
+    attempt = 0
+    while attempt < retries:
+        attempt += 1
```
maybe would be more readable to use `for` 🤷

jeanmarinho529 commented on Tools/build/generate_sbom.py:
```diff
@@ -161,6 +163,23 @@ def get_externals() -> list[str]:
+        except urllib.error.URLError as ex:
+            if attempt == retries:
+                raise ex
+            time.sleep(2**attempt)
```
I think it would be better to make the delay between retries configurable (e.g., base_delay) instead of hardcoding 2 as the base of the exponential backoff. This would give users more control over the retry behavior.

emmatyping replied:
Sure, I don't think it's strictly needed as there are no users; this is a script and the function only has one use in the script.

emmatyping committed:
Reformat function signature

Co-authored-by: Semyon Moroz <donbarbos@proton.me>

emmatyping committed:
Apply suggestions from review

- Add `base_delay` for retries
- use a for loop instead of while loop
- rename retries to max_retries
